Add basic setup to build debian package

joseph-wakeling-sociomantic opened the pull request:
This patch implements the minimum necessary to get a deb package built and uploaded to our repositories.

* beaver ()...v0.1.1(2c6caef) (1 commits)
  > Bugfix: Change TRAVIS_SLUG to TRAVIS_REPO_SLUG

joseph-wakeling-sociomantic commented:
@leandro-lucarella-sociomantic this is an attempt at a minimum setup that should work with Travis.  Let me know if anything looks horrendous with it.

There are a few potential issues:

* the original project already has a `.travis.yml`, it would be good to add ours as an alternative rather than an overwrite

* the original project already has a `docker` directory, it would be good to be able to use a separate directory/docker build script

joseph-wakeling-sociomantic commented on sociomantic-build:
```diff
@@ -0,0 +1,55 @@
+changelog=`mktemp`
+genchangelog "$pkgname" "$pkgversion" "$pkgmaintainer" > "$changelog"
+
+cd -
```
This seems to cause an error when I attempt to build locally:

```sh
+ cd -
/home/my-user-name
+ fpm -s dir -t deb -n libmxnet -v 0.10.0+socio1-trusty --maintainer Sociomantic Tsunami <tsunami@sociomantic.com> --description Lightweight, portable deep learning library --url https://github.com/dmlc/mxnet --vendor Distributed (Deep) Machine Learning Community --license Apache License, Version 2.0 --category libs --depends libc6 --depends libstdc++6 --depends libopenblas-base --deb-changelog /tmp/tmp.YVnJBKdXub --deb-no-default-config-files mxnet/lib/libmxnet.a=/usr/lib/ mxnet/lib/libmxnet.so=/usr/lib/
Invalid package configuration: Cannot chdir to 'mxnet/lib'. Does it exist? {:level=>:error}
```

joseph-wakeling-sociomantic commented on .travis.yml:
```diff
@@ -1,106 +1,27 @@
-sudo: false
+# We will use docker to set up out environment, so don't use any particular
```
Whoops, typo, `out` :-P

joseph-wakeling-sociomantic commented:
Tweaked the build script to avoid the unnecessary `cd -` command:

```diff
diff --git a/.travis.yml b/.travis.yml
index bf440e40..7241342d 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -17,7 +17,7 @@ env:
 
 install: beaver install
 
-script: beaver run ./sociomantic-build
+script: beaver run ./build-deb-package
 
 deploy:
     provider: script
diff --git a/sociomantic-build b/build-deb-package
similarity index 95%
rename from sociomantic-build
rename to build-deb-package
index e62dc559..f13326f7 100755
--- a/sociomantic-build
+++ b/build-deb-package
@@ -38,7 +38,6 @@ make USE_BLAS=openblas USE_OPENCV=0 USE_OPENMP=0
 changelog=`mktemp`
 genchangelog "$pkgname" "$pkgversion" "$pkgmaintainer" > "$changelog"
 
-cd -
 fpm -s dir -t deb -n "$pkgname" -v "$pkgversion" \
     --maintainer "$pkgmaintainer" \
     --description "Lightweight, portable deep learning library" \
@@ -51,5 +50,5 @@ fpm -s dir -t deb -n "$pkgname" -v "$pkgversion" \
     --depends libopenblas-base \
     --deb-changelog "$changelog" \
     --deb-no-default-config-files \
-    mxnet/lib/libmxnet.a=/usr/lib/ \
-    mxnet/lib/libmxnet.so=/usr/lib/
+    ./lib/libmxnet.a=/usr/lib/ \
+    ./lib/libmxnet.so=/usr/lib/
```

joseph-wakeling-sociomantic commented:
... and fixed a typo in `.travis.yml`:

```diff
diff --git a/.travis.yml b/.travis.yml
index 7241342d..ddd7065e 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -1,4 +1,4 @@
-# We will use docker to set up out environment, so don't use any particular
+# We will use docker to set up our environment, so don't use any particular
 # language in Travis itself
 language: generic
```

joseph-wakeling-sociomantic commented on build-deb-package:
```diff
@@ -0,0 +1,54 @@
+# Copyright sociomantic labs GmbH 2017.
+# Distributed under the Boost Software License, Version 1.0.
+# (See accompanying file BOOST_LICENSE_1.0.txt or copy at
+# http://www.boost.org/LICENSE_1_0.txt)
```
I wonder if we could just drop this and implicitly assume its licence to match that of the upstream project (Apache v2.0: https://github.com/sociomantic-tsunami/mxnet/blob/master/LICENSE)

leandro-lucarella-sociomantic replied:
Yeah, let's go with the project's license, I just picked the license we use the most since it was a separate project.

leandro-lucarella-sociomantic commented on build-deb-package:
```diff
@@ -0,0 +1,54 @@
+# if the requested build identifier matches a git tag, use that directly,
+# otherwise use git revision description (note, we implicitly assume that all
+# tags and revision descriptions will begin with `v`)
+versiontag=$(git describe)
```
Why did you stripped the version scheme conversion code from here (compared to https://github.com/sociomantic-tsunami/libmxnet-deb/pull/2/files#diff-b0da275520918e23dd615e2a747528f1R28)? We need to convert the version from SemVer to Debian scheme. You could get the code from [MakD](https://github.com/sociomantic-tsunami/makd/blob/v1.x.x/Makd.mak#L251-L256).

To void duplication, I wonder if it's not worth adding MakD as a submodule just for the packaging building...

leandro-lucarella-sociomantic replied:
Also why calling it `versiontag` and not just version?

joseph-wakeling-sociomantic replied:
Ah, I think I missed the meaning of some of that code.  I meant to just strip out the stuff that uses either `$ref` or $(git describe) depending on whether `$ref` is a tag.  I'll restore.

joseph-wakeling-sociomantic replied:
Actually, I don't understand what you are referring to here.  The code you reference doesn't convert from SemVer to debian versioning, it just checks whether `ref` is a valid tag: if it is, it is used as the version, and if not, the output of `git describe` is used instead.

Is there something else that should be used here instead?

joseph-wakeling-sociomantic replied:
I'm going to copy the equivalent from `libdrizzle` since I assume this is what you are referring to.

leandro-lucarella-sociomantic replied:
Yes, it does convert the version, it changes `-` with `~`. Have you read the comment? Is explained in there :)

joseph-wakeling-sociomantic replied:
@leandro-lucarella-sociomantic are you talking about the version code in libdrizzle or the stuff linked to?  I genuinely don't see the `-` to `~` conversion in the code linked to.

leandro-lucarella-sociomantic commented on build-deb-package:
```diff
@@ -0,0 +1,54 @@
+pkgversion="${pkgversion_number}-${pkgdistro}"
+pkgmaintainer="Sociomantic Tsunami <tsunami@sociomantic.com>"
+
+DEB_HOST_MULTIARCH=x86_64-linux-gnu
```
What is this? I don't see it used anywhere...

joseph-wakeling-sociomantic replied:
Yea, probably a leftover of some copy-paste from elsewhere.  I'll cut, and restore if needed.

joseph-wakeling-sociomantic commented:
I've pushed a fixup patch.  This tweaks the calculation of the package version number in line with what libdrizzle does (though I've retained the `sed`-based stripping of `v` instead of the `cut`-based approach used by libdrizzle).  I've also switched the license and removed the weird `DEB_HOST_MULTIARCH`.

I'll repeat my earlier question: is it possible for us to use an alternative travis config file rather than needing to overwrite upstream's `.travis.yml`?  And how would you recommend structuring the placement of the package-build script, and the docker-related stuff (see questions above)?

joseph-wakeling-sociomantic commented:
Note, the package versions produced by this script are as follows:

* without an explicit tag on the current commit: `0.9.3a+660~g6a3480e9~trusty` (I think the MXNet folks failed to make `v0.10.0` an annotated tag...)

* with an explicit tag: `0.10.0+socio1~trusty` (the tag was `v0.10.0+socio1`)

joseph-wakeling-sociomantic commented:
@leandro-lucarella-sociomantic is there any way to indicate to Travis that it should use `--no-recommend-shallow` for the submodule checkout?

joseph-wakeling-sociomantic commented:
I've pushed a further fixup patch to (hopefully) address the submodule fetch issue and ensure we do a non-shallow clone.

joseph-wakeling-sociomantic commented:
OK, shallow clone works, but the docker setup is borked.  I've pushed a fresh patch that I hope should fix this (including the missing `build` file that docker was looking for).

```diff
diff --git a/.dockerignore b/.dockerignore
index 55c7104b..67179896 100644
--- a/.dockerignore
+++ b/.dockerignore
@@ -1,2 +1,3 @@
 *
 !docker/
+!deb-package/docker
diff --git a/Dockerfile.trusty b/Dockerfile.trusty
index fee67b5d..5a484f55 100644
--- a/Dockerfile.trusty
+++ b/Dockerfile.trusty
@@ -1,3 +1,3 @@
 FROM sociomantictsunami/base:trusty-v2
-COPY docker/ /docker-tmp
+COPY deb-package/docker/ /docker-tmp
 RUN /docker-tmp/build && rm -fr /docker-tmp
diff --git a/Dockerfile.xenial b/Dockerfile.xenial
index 1ce5eea6..ebf6e3d1 100644
--- a/Dockerfile.xenial
+++ b/Dockerfile.xenial
@@ -1,3 +1,3 @@
 FROM sociomantictsunami/base:xenial-v2
-COPY docker/ /docker-tmp
+COPY deb-package/docker/ /docker-tmp
 RUN /docker-tmp/build && rm -fr /docker-tmp
diff --git a/build-deb-package b/deb-package/build
similarity index 100%
rename from build-deb-package
rename to deb-package/build
diff --git a/deb-package/docker/build b/deb-package/docker/build
new file mode 100644
index 00000000..459da7c7
--- /dev/null
+++ b/deb-package/docker/build
@@ -0,0 +1,5 @@
+#!/bin/sh
+set -xeu
+
+apt-get update
+apt-get -y install libblas-dev libopenblas-dev
```

joseph-wakeling-sociomantic commented:
Note, I've also moved everything I can to a `deb-package` subdirectory (reflecting the upstream naming scheme for directories).

joseph-wakeling-sociomantic commented:
Whoops, forgot to tweak the invocation of the build script:

```diff
diff --git a/.travis.yml b/.travis.yml
index c43288b6..febd69ad 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -24,7 +24,7 @@ env:
 
 install: beaver install
 
-script: beaver run ./build-deb-package
+script: beaver run ./deb-package/build
 
 deploy:
     provider: script
```

joseph-wakeling-sociomantic commented:
.... grrrr ....

```diff
diff --git a/deb-package/docker/build b/deb-package/docker/build
old mode 100644
new mode 100755
```

joseph-wakeling-sociomantic commented:
Well, at least it's building now.  This looks worrying in the build log, though:

```
+ pkgname=libmxnet
+ sed s/^v\(.*\)$/\1/
+ sed s/-\([0-9]\+\)-\(g[0-9a-f]\+\)/+\1~\2/
+ git describe --dirty
+ sed s/\(~g[0-9a-f]\+\)-dirty$/-dirty\1/
+ date +%Y%m%d%H%M%S
fatal: No names found, cannot describe anything.
+ sed s/-dirty/~dirty.20170705132800/
+ pkgversion_number=
+ lsb_release -cs
+ pkgdistro=xenial
+ pkgversion=~xenial
```

i.e. the `git describe` call is failing and hence we get no version number.

joseph-wakeling-sociomantic commented:
OK @leandro-lucarella-sociomantic, I think it's time for your input again -- I don't understand why the `git describe` calls are failing ;-)  I'm guessing it would be because travis is doing a shallow checkout of the project repo, and hence is missing the info which allows `git describe` to work?

joseph-wakeling-sociomantic commented:
Adding `--tags` to the `git describe` call makes no difference, it seems.  I can't see anything obviously different about the libdrizzle travis/build setup, so @leandro-lucarella-sociomantic can you advise what might be done to stop this from happening?

joseph-wakeling-sociomantic commented:
OK, I'm brute forcing this with an explicit `git fetch --unshallow --tags`.  Happy to replace with something else in future, but I don't trust `--depth` settings to get the right result.

joseph-wakeling-sociomantic commented:
Woohoo!  Success :-)

@leandro-lucarella-sociomantic if you can re-review this and maybe make some suggestions (especially w.r.t. the `.travis.yml` concerns) I would be very grateful.

joseph-wakeling-sociomantic commented:
I've pushed a further fixup with an alternative to the explicit `git fetch --unshallow --tags` call, just to see if this will also work.

joseph-wakeling-sociomantic commented:
Seems to work, but I'm still minded to do the explicit `git fetch --unshallow` as it makes explicit the requirements rather than just using a magic `depth` value.

(Incidentally, I don't really understand why the bigger depth value does work, since 50 is more than enough to ensure that at least one tag is in the fetched version history....)

nemanja-boric-sociomantic commented on deb-package/build:
```diff
@@ -0,0 +1,52 @@
+# Builds a Debian package for MXNET
+#
+# Copyright sociomantic labs GmbH 2017.
+# Distributed under the Apache License, Version 2.0.
```
Strange that we choose Apache for this? Is there any reason for doing so?

nemanja-boric-sociomantic replied:
Ah, well, the entire project is Apache 2.0 licensed.

joseph-wakeling-sociomantic replied:
@nemanja-boric-sociomantic it's the upstream project license, so we might as well play nice.  @leandro-lucarella-sociomantic already discussed this and approved the choice.

The alternative would be to use Boost and include the Boost license directly in the file (I really don't fancy confusing things by putting the boost license in a separate file in the project root).

nemanja-boric-sociomantic replied:
All clear, thanks!

nemanja-boric-sociomantic commented on deb-package/build:
```diff
@@ -0,0 +1,52 @@
+                    sed 's/-dirty/~dirty.'`date +%Y%m%d%H%M%S`'/'
+                   )
+pkgdistro=$(lsb_release -cs)
+pkgversion="${pkgversion_number}~${pkgdistro}"
```
Just a question: why not `-` here instead of `~`?

joseph-wakeling-sociomantic replied:
Going by typical Ubuntu package versions, ~ is usually used before this kind of version qualifier.  But I'm happy to tweak it however would be preferred.

The main thing I'd like to address is to ensure that any `+socioN` in the semver tag becomes a `-socioN` in the package version.  That's presumably another straightforward `sed` command (`s/+socio/-socio/` should suffice).

Then we wind up with a package version like `0.10.0-socio1~xenial` which is comparable to (for example) the `git` package version `1:2.13.0-0ppa1~ubuntu14.04.1`.

leandro-lucarella-sociomantic replied:
I think there is quite some confusion here on how versions are handled by SemVer and Debian. In SemVer `-` is to mark a "pre-release", in Debian is a separator to differentiate between the upstream version and the debian build (for example, if upstream version v1.2.3 is packaged for the first time, then the debian version will probably be something like 1.2.3-0. Suppose that the pre-install script in the package had a bug, and you want to build another package, then you fix the script in the package and make the version 1.2.3-1. The upstream version is the exact same, is just the "package" version that changes. There is no equivalence for this in SemVer.
Then in SemVer `+` is used to attach build-metadata, which is sort of similar to the previous but not the same, because in terms of version calculation (if a version is greater than another) is a NOP. In Debian the build-metadata is specified by `~`.

Seeing this, it might make more sense to use the `~` to specify the distro in the package version, that is a good point, but having `pkgversion="${pkgversion_number}~${pkgdistro}"` doesn't make a lot of sense IMHO, because this now isn't the changing the package version part but the upstream version part, if there is no `-` in the Debian version (and relying in tagging with `+socio1` might not be reliable).

Also the conversion between `+socioN` to `-socioN` changes the meaning in the version, as the former is metadata that doesn't participate in version calculation while the later isn't.

In any case, what it clear is that this is tricky. So I'm more inclined than before that we should use Makd to make these conversions.

joseph-wakeling-sociomantic replied:
I have just one concern about using makd: it would be nice if our deb-creating solution was at least in principle upstreamable.  `beaver` is just a bunch of convenience stuff for our Travis setup, which we could presumably avoid when upstreaming, but makd seems a bit more intrusive.

leandro-lucarella-sociomantic replied:
> I have just one concern about using makd: it would be nice if our deb-creating solution was at least in principle upstreamable.

I don't agree really, packaging should be done by packagers, not upstream. To have a real debian compatible package there are tons of policies we are not going to comply.

I never knew upstreaming this was in the plans, but I wouldn't go that route.

joseph-wakeling-sociomantic replied:
> I never knew upstreaming this was in the plans, but I wouldn't go that route.

Don't worry, it wasn't a strong plan.  I just noticed that the upstream repo already has quite a bit of packaging-related stuff in there (including a snap-package definition!), so it occurred to me it might be something that would be welcomed.

joseph-wakeling-sociomantic commented:
Reverted the `depth` tweak in favour of the explicit `git fetch --unshallow` and added a further fixup commit to convert `s/+socio/-socio/` when converting the semver tag to the deb package version:

* with an explicit tag of `v0.10.0+socio1` this gives us `0.10.0-socio1~trusty` for the package

* without any explicit tag we get `0.10.0+7~gd019c89d~trusty`

Will that do version-wise? :-)

joseph-wakeling-sociomantic commented on deb-package/build:
```diff
@@ -0,0 +1,53 @@
+pkgname=libmxnet
+pkgversion_number=$(git describe --tags --dirty |
+                    sed 's/^v\(.*\)$/\1/' |
+					sed 's/+socio/-socio/' |
```
... whoops, tags need replacing with spaces :-P  I'll do this when squashing these patches.

joseph-wakeling-sociomantic commented:
> Will that do version-wise?

What I'm mulling over is whether I should explicitly special-case `socio` in this context, or whether I should just do a general `s/+/-/` (which e.g. translates `0.10.0+1` => `0.10.0-1`, which also matches my experience of deb package versioning for upstream + patches).

joseph-wakeling-sociomantic commented on deb-package/build:
```diff
@@ -0,0 +1,53 @@
+    echo "$1 ($2) $(lsb_release -sc); urgency=low"
+    echo
+    prevtag=$(git describe --tags --abbrev=0 HEAD^)
+    git log --date=short --format=" * %s (%h, $cd)" "$prevtag"..HEAD |
```
This is generating an error in the build log:

```
./deb-package/build: 29: ./deb-package/build: cd: parameter not set
```

It looks like a typo: it should probably be `%cd`.

leandro-lucarella-sociomantic replied:
Yes.

joseph-wakeling-sociomantic commented:
Pushed a few more fixes: one for the typo `$cd` => `%cd`, the other to generalize the `+socio` => `-socio` replacement.

@leandro-lucarella-sociomantic I promise I'm going to pause the slew of fixup patches now, so you can have a new review of this ;-)

leandro-lucarella-sociomantic commented on .travis.yml:
```diff
@@ -1,106 +1,34 @@
-      - pdl
-      - cpanminus
-      - swig
-      - libgraphviz-perl
```
BTW, shouldn't all these go to the Dockerfile?

joseph-wakeling-sociomantic replied:
... wha?  I don't get what you are referring to.  All the deleted lines are from the upstream project's `.travis.yml`, which I am replacing.

I'm not responsible for how they choose to declare their project dependencies ;-)

leandro-lucarella-sociomantic replied:
No, but you need to build it, so you'll need the dependencies. I don't understand how it's building without them, either they are optional and we are building a limited version, or they are installnig a lot of unnecessary dependencies, or I'm missing something.

joseph-wakeling-sociomantic replied:
@leandro-lucarella-sociomantic no, we don't need those dependencies.  All of this stuff is for things not in the underlying C library.  How do you think we were able to build `libmxnet` all this time without them? ;-)

leandro-lucarella-sociomantic commented on deb-package/build:
```diff
@@ -10,7 +10,7 @@ set -xeu
 # Set up package
 pkgname=libmxnet
-pkgversion_number=$(git describe --dirty |
+pkgversion_number=$(git describe --tags --dirty |
```
Please don't add this, tags should be always annotated.

joseph-wakeling-sociomantic replied:
I added this because not all upstream tags are annotated.  Obviously we can control the tags we apply, but I was getting frustrated at the builds of untagged commits coming out with versions like `0.9.3a+666` instead of `0.10.0+3`.

But I'll delete if you strongly prefer.

leandro-lucarella-sociomantic replied:
Yes, we will only care about tags we make and using annotated tags can introduce problems as most our infra assumes tags are annotated.

leandro-lucarella-sociomantic commented on .travis.yml:
```diff
@@ -9,11 +9,11 @@ services:
 
 # Disable automatic submodule fetching ...
 git:
+    depth: 9999999
```
Mmmm... I think having such a complete clone just to get a complete changelog is a bit too much, I would just remove the changelog (or let it generate only what we have). Again using makd should help here, otherwise we are duplicating work.

joseph-wakeling-sociomantic replied:
Can you point me to an example of using makd in this context?  (Note that the project already has its own makefiles.)

leandro-lucarella-sociomantic replied:
`make -f makd/Makd SRC=. pkg` basically.

leandro-lucarella-sociomantic replied:
Actually since the project has a `src/` directory, `SRC=.` is not even necessary :)

Then of course you need to add the package description files in `pkg/` as usual.

joseph-wakeling-sociomantic replied:
OK, I'll see if makd can help w.r.t. versioning and the depth issue.

leandro-lucarella-sociomantic commented:
Also, there is so much back and forth in the commits now that it's making all harder to review, I recommend squashing now before moving forward.

joseph-wakeling-sociomantic commented:
OK, here's an alternative patch using makd as suggested.  (Thanks _very_ much for pressing on this: it does make for a much simpler setup.)

```diff
diff --git a/.dockerignore b/.dockerignore
index 67179896..05e4a789 100644
--- a/.dockerignore
+++ b/.dockerignore
@@ -1,3 +1,3 @@
 *
 !docker/
-!deb-package/docker
+!pkg/docker
diff --git a/.gitmodules b/.gitmodules
index 48620bc7..b7727632 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -17,3 +17,6 @@
 [submodule "beaver"]
        path = beaver
        url = https://github.com/sociomantic-tsunami/beaver.git
+[submodule "makd"]
+       path = makd
+       url = https://github.com/sociomantic-tsunami/makd.git
diff --git a/.travis.yml b/.travis.yml
index a9d7b0d2..8649c11b 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -26,7 +26,9 @@ env:
 
 install: beaver install
 
-script: beaver run ./deb-package/build
+script:
+    - make USE_BLAS=openblas USE_OPENCV=0 USE_OPENMP=0
+    - make -f makd/Makd.mak pkg
 
 deploy:
     provider: script
diff --git a/Dockerfile.trusty b/Dockerfile.trusty
index 5a484f55..5eafcf8d 100644
--- a/Dockerfile.trusty
+++ b/Dockerfile.trusty
@@ -1,3 +1,3 @@
 FROM sociomantictsunami/base:trusty-v2
-COPY deb-package/docker/ /docker-tmp
+COPY pkg/docker/ /docker-tmp
 RUN /docker-tmp/build && rm -fr /docker-tmp
diff --git a/Dockerfile.xenial b/Dockerfile.xenial
index ebf6e3d1..82a77198 100644
--- a/Dockerfile.xenial
+++ b/Dockerfile.xenial
@@ -1,3 +1,3 @@
 FROM sociomantictsunami/base:xenial-v2
-COPY deb-package/docker/ /docker-tmp
+COPY pkg/docker/ /docker-tmp
 RUN /docker-tmp/build && rm -fr /docker-tmp
diff --git a/deb-package/build b/deb-package/build
deleted file mode 100755
index cdd62659..00000000
--- a/deb-package/build
+++ /dev/null
@@ -1,53 +0,0 @@
-#!/bin/sh
-#
-# Builds a Debian package for MXNET
-#
-# Copyright sociomantic labs GmbH 2017.
-# Distributed under the Apache License, Version 2.0.
-# (See accompanying LICENSE file or copy at
-# https://www.apache.org/licenses/LICENSE-2.0.txt)
-set -xeu
-
-# Set up package
-pkgname=libmxnet
-pkgversion_number=$(git describe --tags --dirty |
-                    sed 's/^v\(.*\)$/\1/' |
-                    sed 's/\+/\-/' |
-                    sed 's/-\([0-9]\+\)-\(g[0-9a-f]\+\)/+\1~\2/' |
-                    sed 's/\(~g[0-9a-f]\+\)-dirty$/-dirty\1/' |
-                    sed 's/-dirty/~dirty.'`date +%Y%m%d%H%M%S`'/'
-                   )
-pkgdistro=$(lsb_release -cs)
-pkgversion="${pkgversion_number}~${pkgdistro}"
-pkgmaintainer="Sociomantic Tsunami <tsunami@sociomantic.com>"
-
-genchangelog()
-{
-    echo "$1 ($2) $(lsb_release -sc); urgency=low"
-    echo
-    prevtag=$(git describe --tags --abbrev=0 HEAD^)
-    git log --date=short --format=" * %s (%h, %cd)" "$prevtag"..HEAD |
-            fold --spaces --width 76 | sed 's/^\([^ ]\+\)/    \1/'
-    echo
-    echo " -- $3  `LANG=C date -R`"
-}
-
-make USE_BLAS=openblas USE_OPENCV=0 USE_OPENMP=0
-
-changelog=`mktemp`
-genchangelog "$pkgname" "$pkgversion" "$pkgmaintainer" > "$changelog"
-
-fpm -s dir -t deb -n "$pkgname" -v "$pkgversion" \
-    --maintainer "$pkgmaintainer" \
-    --description "Lightweight, portable deep learning library" \
-    --url 'https://github.com/dmlc/mxnet' \
-    --vendor 'Distributed (Deep) Machine Learning Community' \
-    --license 'Apache License, Version 2.0' \
-    --category libs \
-    --depends libc6 \
-    --depends libstdc++6 \
-    --depends libopenblas-base \
-    --deb-changelog "$changelog" \
-    --deb-no-default-config-files \
-    ./lib/libmxnet.a=/usr/lib/ \
-    ./lib/libmxnet.so=/usr/lib/
diff --git a/makd b/makd
new file mode 160000
index 00000000..8711da22
--- /dev/null
+++ b/makd
@@ -0,0 +1 @@
+Subproject commit 8711da2276dec795faa2a718c44ef5b18c859c0b
diff --git a/deb-package/docker/build b/pkg/docker/build
similarity index 100%
rename from deb-package/docker/build
rename to pkg/docker/build
diff --git a/pkg/libmxnet.pkg b/pkg/libmxnet.pkg
new file mode 100644
index 00000000..5db07e2c
--- /dev/null
+++ b/pkg/libmxnet.pkg
@@ -0,0 +1,14 @@
+libs = [VAR.shortname + '.a', VAR.shortname + '.so']
+
+OPTS = dict(
+    name = VAR.fullname,
+    description = 'Lightweight, portable deep learning library',
+    url = 'https://github.com/dmlc/mxnet',
+    maintainer = 'Sociomantic Tsunami <tsunami@sociomantic.com>',
+    vendor = 'Distributed (Deep) Machine Learning Community',
+    license = 'Apache License, Version 2.0',
+    category = 'libs',
+    depends = ['libc6', 'libstdc++6', 'libopenblas-base'],
+)
+
+ARGS = FUN.mapfiles('./lib', '/usr/lib', libs)
```

joseph-wakeling-sociomantic commented:
Added a few small tweaks to the packaging setup:

```diff
diff --git a/pkg/libmxnet.pkg b/pkg/libmxnet.pkg
index 5db07e2c..4222a54f 100644
--- a/pkg/libmxnet.pkg
+++ b/pkg/libmxnet.pkg
@@ -8,7 +8,12 @@ OPTS = dict(
     vendor = 'Distributed (Deep) Machine Learning Community',
     license = 'Apache License, Version 2.0',
     category = 'libs',
+    provides = VAR.shortname,
+    conflicts = VAR.shortname,
     depends = ['libc6', 'libstdc++6', 'libopenblas-base'],
 )
 
-ARGS = FUN.mapfiles('./lib', '/usr/lib', libs)
+ARGS = FUN.mapfiles('./lib', '/usr/lib', libs, append_suffix=False) + [
+    'LICENSE=/usr/share/doc/' + VAR.fullname + '/copyright',
+    'README.md=/usr/share/doc/' + VAR.fullname + '/',
+]
```

joseph-wakeling-sociomantic commented:
@leandro-lucarella-sociomantic assuming this passes tests, this is ready for your re-review :-)

A favour to ask: assuming you approve, can you hold off merging it until I've written a slightly more in-depth commit message?  It would be good to summarize in a little more detail what has been done and why, just for the record.

joseph-wakeling-sociomantic commented:
Hmmm, this build failure is a bit of a mystery:

```
[0K$ make USE_BLAS=openblas USE_OPENCV=0 USE_OPENMP=0
g++ -std=c++11 -c -DMSHADOW_FORCE_STREAM -Wall -Wsign-compare -O3 -DNDEBUG=1 -I/home/travis/build/sociomantic-tsunami/mxnet/mshadow/ -I/home/travis/build/sociomantic-tsunami/mxnet/dmlc-core/include -fPIC -I/home/travis/build/sociomantic-tsunami/mxnet/nnvm/include -Iinclude -funroll-loops -Wno-unused-variable -Wno-unused-parameter -Wno-unknown-pragmas -Wno-unused-local-typedefs -msse3 -DMSHADOW_USE_CUDA=0 -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_RABIT_PS=0 -DMSHADOW_DIST_PS=0 -DMSHADOW_USE_PASCAL=0 -DMXNET_USE_OPENCV=0  -DMXNET_USE_NVRTC=0 -MMD -c src/operator/contrib/count_sketch.cc -o build/src/operator/contrib/count_sketch.o
In file included from /home/travis/build/sociomantic-tsunami/mxnet/mshadow/mshadow/tensor.h:16:0,
                 from include/mxnet/./base.h:13,
                 from include/mxnet/operator.h:19,
                 from src/operator/contrib/./count_sketch-inl.h:11,
                 from src/operator/contrib/count_sketch.cc:7:
/home/travis/build/sociomantic-tsunami/mxnet/mshadow/mshadow/./base.h:136:23: fatal error: cblas.h: No such file or directory
     #include <cblas.h>
                       ^
compilation terminated.
make: *** [build/src/operator/contrib/count_sketch.o] Error 1
```

... because `libblas-dev` is definitely installed when the docker image is built (see the build log).

@leandro-lucarella-sociomantic any idea what could be wrong here?  It's odd, because we never saw this build error before.

leandro-lucarella-sociomantic commented on .dockerignore:
```diff
@@ -0,0 +1,3 @@
+*
+!docker/
+!pkg/docker
```
What is this for? Keep it simple please and just put whaever is needed to build the image in `docker/` that's the only purpose of that directory :)

joseph-wakeling-sociomantic replied:
I didn't want to interfere with upstream's `docker` directory, so created a `pkg/docker` directory instead.

Previously it was `deb-package/docker`, and this seemed to work just fine.

leandro-lucarella-sociomantic replied:
Oh, there is an upstream `docker/` directory? Then please ignore it via this `.dockerignore`. If there was no `.dockerignore` before, then the upstream dockerfile will be broken anyway if they didn't ignore the rest.

leandro-lucarella-sociomantic replied:
This is still pending...

joseph-wakeling-sociomantic replied:
I'm not sure I follow what you want here.  There is an upstream `docker` directory which I preferred to leave alone, so I put the docker `build` file in `pkg/docker/build` instead of `docker/build`.

That means that I needed to ignore both the upstream `docker/` dir and the added `pkg/docker` dir.

Is that correct, or would you prefer me to do something else?

joseph-wakeling-sociomantic replied:
I ask because AFAICS the `.dockerignore` file correctly ignores both `docker/` and `pkg/docker`.

leandro-lucarella-sociomantic replied:
> That means that I needed to ignore both the upstream docker/ dir and the added pkg/docker dir.

This is the part that you got wrong :)

You are **un**ignoring those directories. This means, when you run `docker build`, the docker client will only pass `docker/` and `pkg/docker/` contents to the docker daemon to build the image. What I'm saying is you should only pass `pkg/docker/`, since we want to ignore the original docker stuff in the upstream repo.

To be more clear, you only need the stuff in `pkg/docker/` when building the docker image, hence you should only have this in the `.dockerignore` file:
```
*             # Ignore everything
!pkg/docker/  # but don't ignore the pkg/docker/ directory
```
(I'm not sure if comments are valid there, only including them for clarification)

joseph-wakeling-sociomantic replied:
Ah, gotcha!  I must have misunderstood the syntax;  that does clarify why `*` is the first entry....

I'll fix.

joseph-wakeling-sociomantic commented:
@leandro-lucarella-sociomantic can I ping here on what's potentially going wrong with the docker build and the `libblas-dev` and `fpm` dependencies ... ?

leandro-lucarella-sociomantic commented:
> A favour to ask: assuming you approve, can you hold off merging it until I've written a slightly more in-depth commit message? It would be good to summarize in a little more detail what has been done and why, just for the record.

Yes, in general I use the following scheme to indicated this, assign yourself the PR and just make me a reviewer. Then when I approve you are free to merge whenever is more convenient to you. If you feel you need another review, just poke me again, but as long as you are assigned, I won't ever merge, you merge :)

leandro-lucarella-sociomantic commented:
> @leandro-lucarella-sociomantic can I ping here on what's potentially going wrong with the docker build and the libblas-dev and fpm dependencies ... ?

I'll have a look now.

leandro-lucarella-sociomantic commented on .travis.yml:
```diff
@@ -1,106 +1,38 @@
-    on_success: change
-    on_failure: always
-  #slack: dmlc:NmroCzntCiWOuxUZpii40USd
+    - make USE_BLAS=openblas USE_OPENCV=0 USE_OPENMP=0
```
`beaver make ...` in both. This is why you get all the errors, it's not running inside the docker container :)

joseph-wakeling-sociomantic commented:
Corrected to use `beaver make`:

```diff
diff --git a/.travis.yml b/.travis.yml
index 8649c11b..02c6ac5a 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -27,8 +27,8 @@ env:
 install: beaver install
 
 script:
-    - make USE_BLAS=openblas USE_OPENCV=0 USE_OPENMP=0
-    - make -f makd/Makd.mak pkg
+    - beaver make USE_BLAS=openblas USE_OPENCV=0 USE_OPENMP=0
+    - beaver make -f makd/Makd.mak pkg
 
 deploy:
     provider: script
```

joseph-wakeling-sociomantic commented:
... well, we're getting closer:

```
+./beaver/bin/beaver run make -rj2 -f makd/Makd.mak pkg
+docker run -ti --rm -v /home/travis:/home/travis -v /home/travis/build/sociomantic-tsunami/mxnet:/home/travis/build/sociomantic-tsunami/mxnet -w /home/travis/build/sociomantic-tsunami/mxnet -u 2000 -e USER -e HOME -e LANG -e LC_ALL -e DEBIAN_FRONTEND -e RAILS_ENV -e RACK_ENV -e MERB_ENV -e JRUBY_OPTS -e JAVA_HOME -e CI -e CONTINUOUS_INTEGRATION -e TRAVIS_OS_NAME -e TRAVIS_STACK_LANGUAGES -e TRAVIS_BRANCH -e TRAVIS_SUDO -e TRAVIS_PRE_CHEF_BOOTSTRAP_TIME -e TRAVIS_PULL_REQUEST_BRANCH -e TRAVIS_PULL_REQUEST_SLUG -e TRAVIS_COMMIT_MESSAGE -e TRAVIS_STACK_NAME -e TRAVIS_STACK_JOB_BOARD_REGISTER -e TRAVIS_REPO_SLUG -e TRAVIS_BUILD_ID -e TRAVIS_COMMIT_RANGE -e TRAVIS -e TRAVIS_ALLOW_FAILURE -e TRAVIS_EVENT_TYPE -e TRAVIS_STACK_FEATURES -e TRAVIS_BUILD_DIR -e TRAVIS_BUILD_NUMBER -e TRAVIS_SECURE_ENV_VARS -e TRAVIS_TAG -e TRAVIS_JOB_ID -e TRAVIS_COMMIT -e TRAVIS_PULL_REQUEST -e TRAVIS_STACK_TIMESTAMP -e TRAVIS_JOB_NUMBER -e TRAVIS_UID -e TRAVIS_PULL_REQUEST_SHA -e TRAVIS_LANGUAGE -e TRAVIS_STACK_NODE_ATTRIBUTES -e TRAVIS_TEST_RESULT -e DIST -e BINTRAY_USER -e BINTRAY_KEY beaver make -rj2 -f makd/Makd.mak pkg
/home/travis/build/sociomantic-tsunami/mxnet/makd/git-deblog: warning: not using tag v0.10.0 because is not annotated
/home/travis/build/sociomantic-tsunami/mxnet/makd/git-deblog: warning: not using tag v0.10.0-rc0 because is not annotated
/home/travis/build/sociomantic-tsunami/mxnet/makd/git-deblog: warning: not using tag v0.9.5 because is not annotated
/home/travis/build/sociomantic-tsunami/mxnet/makd/git-deblog: warning: not using tag v0.8.0 because is not annotated
/home/travis/build/sociomantic-tsunami/mxnet/makd/git-deblog: warning: not using tag nightly because is not annotated
Traceback (most recent call last):
  File "/home/travis/build/sociomantic-tsunami/mxnet/makd/git-deblog", line 259, in <module>
    main()
  File "/home/travis/build/sociomantic-tsunami/mxnet/makd/git-deblog", line 60, in main
    print('  * {} ({})'.format(obj.message, obj.hash[:7]))
UnicodeEncodeError: 'ascii' codec can't encode character '\u2026' in position 73: ordinal not in range(128)
make: *** [/home/travis/build/sociomantic-tsunami/mxnet/build/devel/tmp/pkg-libmxnet.stamp] Error 1
```

@leandro-lucarella-sociomantic any thoughts ... ? :-)

leandro-lucarella-sociomantic commented:
Oh, crap. unicode errors are the worst! I guess there is some invalid UTF in the commit messages...

leandro-lucarella-sociomantic commented:
Mmm, I can't even reproduce it locally

joseph-wakeling-sociomantic commented:
> Oh, crap. unicode errors are the worst! I guess there is some invalid UTF in the commit messages...

Alas, we cannot control these ;-)  I note that the Travis log seems to contain a lot of invalid characters _anyway_, though.  Is travis set up to log in UTF8, or in ascii?

leandro-lucarella-sociomantic commented:
I assume UTF-8, it might be a docker thing too. I tried to run it using docker locally but I get an issue about `git author` not being set before other errors. I have to go now, but please remind me about this on monday :)

joseph-wakeling-sociomantic commented:
Will do.  And no worries: if this is a nasty issue, it's a nasty issue (and thanks for looking at it).

My suspicion about ascii versus UTF8 was sparked by this:

```
UnicodeEncodeError: 'ascii' codec can't encode character '\u2026' in position 73: ordinal not in range(128)
```

Note, U2026 is the `...` 3-dot ellipsis character (which I mock up here with 3 actual dots).  Please tell me someone wasn't writng their commit messages in LibreOffice or Word? :-P

joseph-wakeling-sociomantic commented:
> please remind me about this on monday :)

... reminding :-)

leandro-lucarella-sociomantic commented:
OK, I could finally reproduce the problem.

leandro-lucarella-sociomantic commented:
This beaver PR should fix it: https://github.com/sociomantic-tsunami/beaver/pull/12

leandro-lucarella-sociomantic commented:
I'm trying a workaround that you could use until a new beaver release is out in #2.

joseph-wakeling-sociomantic commented:
Thanks very much @leandro-lucarella-sociomantic!

leandro-lucarella-sociomantic commented:
The workaround worked!

joseph-wakeling-sociomantic commented:
> The workaround worked!

Great news, thank you!  Shall I wait on a new `beaver` version and update here once that's done?

leandro-lucarella-sociomantic commented:
Depends on how eager you are to get this working...

leandro-lucarella-sociomantic commented on .travis.yml:
```diff
@@ -1,106 +1,38 @@
-osx_image: xcode8
+# ... in order to ensure we do a non-shallow submodule fetch
+before_install:
+    - git fetch --unshallow --tags
```
This was never been clear to me, why do you need to do this `fetch --unshallow --tags`?

joseph-wakeling-sociomantic replied:
IIRC we were getting some nasty issues w.r.t. either commits or tags not being available.  I can try removing this once the `beaver` fix lands.

joseph-wakeling-sociomantic replied:
The "nasty issues" were that `git describe` would fail.  IIRC the `--unshallow` was to make sure that we could do the `git describe HEAD^` call used when generating the changelog, while the `--tags` was to make sure we had all upstream tags (which might not be such a big deal, but had some irritations).

leandro-lucarella-sociomantic replied:
And this still happens with makd? Because this was solely needed for the changelog, right?

joseph-wakeling-sociomantic replied:
I'll verify.  Probably the best way to handle the changelog issue is to tag the upstream `v0.10.0` release wth `v0.10.0.s0` (i.e. our version, vanilla copy of upstream) and to tag the patched version with `v0.10.0.s1` (meaning we have two well-defined annotated tags).

Upstream, it seems, is lax when it comes to decent tagging habits ;-)

joseph-wakeling-sociomantic commented:
Updated to use latest beaver master.  Let's confirm if this builds correctly ...

joseph-wakeling-sociomantic commented:
OK, success!  @leandro-lucarella-sociomantic note these lines of output when building the package:

```
/home/travis/build/sociomantic-tsunami/mxnet/makd/git-deblog: warning: not using tag v0.10.0 because is not annotated
/home/travis/build/sociomantic-tsunami/mxnet/makd/git-deblog: warning: not using tag v0.10.0-rc0 because is not annotated
/home/travis/build/sociomantic-tsunami/mxnet/makd/git-deblog: warning: not using tag v0.9.5 because is not annotated
/home/travis/build/sociomantic-tsunami/mxnet/makd/git-deblog: warning: not using tag v0.8.0 because is not annotated
/home/travis/build/sociomantic-tsunami/mxnet/makd/git-deblog: warning: not using tag nightly because is not annotated
```

I'll now try removing the `git fetch --unshallow --tags` line and see how we do.

joseph-wakeling-sociomantic commented:
OK, looks like we can drop the `--unshallow` stuff.  I've pushed a follow-up `fixup!` patch to see if we can also limit the unshallow submodule fetch to just the one problematic submodule.

joseph-wakeling-sociomantic commented:
OK, looks good in terms of reducing the amount of unshallow fetches we do.  @leandro-lucarella-sociomantic I'm going to wait for beaver 0.1.2 and then squash everything together for final review.

In the meanwhile, let me know if you have any more feedback on the changeset here.

joseph-wakeling-sociomantic commented:
Squashed and pushed, now using beaver 0.1.2 (actually, no change...) and with an improved commit message.

After this is validated, I think I will try an experimental fixup patch to see if we can improve the situation with `cub` beyond the current forced unshallow fetch.  But probably this can be considered complete as-is.

joseph-wakeling-sociomantic commented:
Experimental fixup patch pushed.  Let's see if this works ...

joseph-wakeling-sociomantic commented:
Fixup patch seems to work, so I've squashed again.  @leandro-lucarella-sociomantic this is ready for your final review :-)

joseph-wakeling-sociomantic commented:
> Besides the pending comment, LGTM.

I'm not sure I follow what is pending here (see my response to the comment itself).  Can you clarify?

joseph-wakeling-sociomantic commented:
Fixed `.dockerignore` to only un-ignore `pkg/docker`:

```diff
diff --git a/.dockerignore b/.dockerignore
index 05e4a789..7d4c14b4 100644
--- a/.dockerignore
+++ b/.dockerignore
@@ -1,3 +1,2 @@
 *
-!docker/
 !pkg/docker
```

leandro-lucarella-sociomantic commented on .dockerignore:
```diff
@@ -0,0 +1,2 @@
+*
+!pkg/docker
```
I think is more correct to include the trailing `/` (`!pkg/docker/`) to indicate it's a directory.

joseph-wakeling-sociomantic commented:
Nit has been picked:

```diff
diff --git a/.dockerignore b/.dockerignore
index 7d4c14b4..ee10b6f4 100644
--- a/.dockerignore
+++ b/.dockerignore
@@ -1,2 +1,2 @@
 *
-!pkg/docker
+!pkg/docker/
```

leandro-lucarella-sociomantic commented:
Feel free to automerge when test are done.
